### PR TITLE
add _ENFORCE_ONLY_CORE_HEADERS

### LINKED ISSUE #311: STL: Should we make it easy to limit the STL to core features, or block it entirely?
Originally, the STL was monolithic - including any STL header could drag in an unbounded number of headers and runtime support machinery. Recently, we've split the monolith into "core" and "non-core" subsets. Core headers drag in only other core headers, and don't drag in any runtime support. For example:

https://github.com/microsoft/STL/blob/580e61a5f5536016f521cc1b2ca636eadc69bd30/stl/inc/type_traits#L1-L9

This allows customers to use `<type_traits>` and similar machinery (instead of reimplementing it themselves), while avoiding things like `<vector>` that want to allocate memory.

However, we currently have no supported mechanism for a customer to *enforce* that only core STL headers are included. This seems like a reasonable request that could encourage partial use of the STL. We need to decide how this should work (I believe we need to choose a macro name, and make `<yvals.h>` emit a nicely worded error message; also see #209).

We could also add a mechanism to completely block the STL (presumably, a macro that `<yvals_core.h>` checks for). The error message for that could recommend blocking only non-core features.

Finally, we need to decide whether a core-only macro counts as a supported mode; if so, we should consider defining our feature-test macros accordingly (which would require some thought).

### PR DESCRIPTION
**Pull request by fsb4000**
Fixes #311

These are probably terrible error messages.
Feel free to suggest a better :) 

As far as I remember, I already asked. But I still want to clarify. We can't create a failing test. They are only available in the LLVM test suite, right?

### CODE REVIEW

**Comment by miscco** [comment]
We are using lit to run the test. Did you try a `test.compile.fail.cpp`

**Comment by fsb4000** [comment]
>Did you try a test.compile.fail.cpp

no.

Can I have 2 failing tests in 1 folder?

One for _FORBID_ALL_STL_HEADERS, one for _ENFORCE_ONLY_CORE_HEADERS?

File: stl/inc/yvals.h [hunk 1]
```diff
@@ -7,6 +7,12 @@
 #ifndef _YVALS
 #define _YVALS
 #include <yvals_core.h>
+
+#ifdef _ENFORCE_ONLY_CORE_HEADERS
+static_assert(false, "error STL2001: You tried to include a non-core C++ standard header file, although the mode is "
+                     "set in which only core C++ standard header files are allowed");
```
Comment (miscco) [inline stl/inc/yvals.h:13]: Maybe rephrase it a bit
```suggestion
static_assert(false, "error STL2001: You tried to include a non-core C++ standard header file. "
                     "However, with `_ENFORCE_ONLY_CORE_HEADERS` only core C++ standard header files are allowed");
```

File: stl/inc/yvals_core.h [hunk 2]
```diff
@@ -364,6 +364,12 @@
 
 #include <vcruntime.h>
 
+#ifdef _FORBID_ALL_STL_HEADERS
+static_assert(false,
+    "error STL2000: You tried to include a C++ standard header file, although the mode is set in which all C++ "
+    "standard header files are forbidden. Perhaps you meant _ENFORCE_ONLY_CORE_HEADERS?");
```
Comment (miscco) [inline stl/inc/yvals_core.h:370]:
```suggestion
static_assert(false,
    "error STL2000: You tried to include a C++ standard header file. "
    "However, with _FORBID_ALL_STL_HEADERS all C++ standard header files are forbidden. "
    "Did you mean _ENFORCE_ONLY_CORE_HEADERS?");
```

File: stl/inc/yvals.h (hunk 1 above)
Comment (miscco) [reply stl/inc/yvals.h:13]: I am not really married to although / however, but I believe it is important to mention the flag that has been defined

**Comment by miscco** [comment]
I do not know, but you can always just add two folders i guess

File: tests/std/tests/GH_000311_enforce_core_headers/test.compile.cpp [hunk 3]
```diff
@@ -0,0 +1,9 @@
+// Copyright (c) Microsoft Corporation.
```
Comment (miscco) [inline tests/std/tests/GH_000311_enforce_core_headers/test.compile.cpp:1]: The name of the file must be `test.compile.fail.cpp`

File: tests/std/tests/GH_000311_enforce_core_headers/test.compile.cpp [hunk 4]
```diff
@@ -0,0 +1,9 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#define _CORE_HEADERS_ONLY
```
Comment (miscco) [inline tests/std/tests/GH_000311_enforce_core_headers/test.compile.cpp:4]: Why are there two defines?

Comment (fsb4000) [reply tests/std/tests/GH_000311_enforce_core_headers/test.compile.cpp:4]: This is still a test that passes :)

I will see how they work with failing tests in LLVM.

Comment (fsb4000) [reply tests/std/tests/GH_000311_enforce_core_headers/test.compile.cpp:4]: https://github.com/microsoft/STL/blob/bd7adb4a932725f60ba096580c415616486ab64c/stl/inc/__msvc_all_public_headers.hpp#L76-L80

**Comment by fsb4000** [comment]
No. It doesn't work.

When I have `test.compile.fail.cpp`, I get:
```
python tests\utils\stl-lit\stl-lit.py ..\..\..\tests\std\tests\GH_000311_enforce_core_headers
stl-lit.py: C:\Dev\STL\llvm-project\llvm\utils\lit\lit\discovery.py:267: warning: input '..\\..\\..\\tests\\std\\tests\\GH_000311_enforce_core_headers' contained no tests
error: did not discover any tests for provided path(s)
```

**Comment by AdamBucior** [comment]
We need to verify that all non-core headers include `yvals.h` instead of `yvals_core.h`. https://github.com/microsoft/STL/blob/bd7adb4a932725f60ba096580c415616486ab64c/stl/inc/algorithm#L9

File: stl/inc/yvals.h [hunk 5]
```diff
@@ -7,6 +7,12 @@
 #ifndef _YVALS
 #define _YVALS
 #include <yvals_core.h>
+
+#ifdef _ENFORCE_ONLY_CORE_HEADERS
+static_assert(false, "error STL2001: You tried to include a non-core C++ standard header file. "
```
Comment (AdamBucior) [inline stl/inc/yvals.h:12]: What about "Tried to include a non-core C++ Standard Library header file with _ENFORCE_ONLY_CORE_HEADERS defined". Sounds more errory to me

File: stl/inc/yvals_core.h [hunk 6]
```diff
@@ -364,6 +364,12 @@
 
 #include <vcruntime.h>
 
+#ifdef _FORBID_ALL_STL_HEADERS
+static_assert(false, "error STL2000: You tried to include a C++ standard header file. "
```
Comment (AdamBucior) [inline stl/inc/yvals_core.h:368]: Similarly "Tried to include a C++ Standard Library header file with _FORBID_ALL_STL_HEADERS defined". The "did you mean _ENFORCE_ONLY_CORE_HEADERS" seems totaly unnecessary. It's not like these two macros are similar enough to be easily mistaken with each other, they literally do what they say.

Comment (fsb4000) [reply stl/inc/yvals_core.h:368]: I added because of this:

https://github.com/microsoft/STL/issues/311

>We could also add a mechanism to completely block the STL (presumably, a macro that <yvals_core.h> checks for). The error message for that could recommend blocking only non-core features.

But I don't know. Maybe it's not needed...

**Comment by fsb4000** [comment]
@miscco I found two links:

https://github.com/microsoft/STL/blob/bd7adb4a932725f60ba096580c415616486ab64c/tests/std/lit.site.cfg.in#L15

https://github.com/llvm/llvm-project/blob/4e52a04833fb352090498d6d1c013a2c61d75e53/libcxx/test/configs/legacy.cfg.in#L45

So with default configuration we can't create a failing test. But I think it's possible with a custom python script...

**Comment by StephanTLavavej** [comment]
We made a decision on #311 in the weekly maintainer meeting - let's keep `_ENFORCE_ONLY_CORE_HEADERS` but drop the `_FORBID_ALL_STL_HEADERS` functionality, since we haven't found a compelling user scenario for that.

File: tests/std/tests/GH_000311_enforce_core_headers/test.compile.pass.cpp [hunk 7]
```diff
@@ -0,0 +1,9 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#define _CORE_HEADERS_ONLY
+#define _ENFORCE_ONLY_CORE_HEADERS
```
Comment (StephanTLavavej) [inline tests/std/tests/GH_000311_enforce_core_headers/test.compile.pass.cpp:5]: I think we could simply fuse this into `GH_001411_core_headers`. That test also uses the `usual_matrix` and defines `_CORE_HEADERS_ONLY`, so it'll provide the same coverage.

Rationale: Every added test directory incurs a minor cost. It's okay to add new test directories for new features and bugfixes - we definitely went overboard in the past by cramming tons of different tests into a single file (I can't remember what I was thinking with `Dev11_0000000_user_defined_literals` :joy_cat:) - but when we can slightly extend an existing test with no loss of coverage, it's simpler and faster to do that.

File: stl/inc/yvals.h [hunk 8]
```diff
@@ -7,6 +7,11 @@
 #ifndef _YVALS
 #define _YVALS
 #include <yvals_core.h>
+
+#ifdef _ENFORCE_ONLY_CORE_HEADERS
+#error STL1005: Tried to include a non-core C++ Standard Library header file with _ENFORCE_ONLY_CORE_HEADERS defined.
```
Comment (StephanTLavavej) [inline stl/inc/yvals.h:12]: We should use the new `_EMIT_STL_ERROR` machinery added by #2897.

File: stl/inc/yvals.h [hunk 9]
```diff
@@ -7,6 +7,11 @@
 #ifndef _YVALS
 #define _YVALS
 #include <yvals_core.h>
+
+#ifdef _ENFORCE_ONLY_CORE_HEADERS
```
Comment (StephanTLavavej) [inline stl/inc/yvals.h:11]: I think this should be checked within `_STL_COMPILER_PREPROCESSOR` - we want to expose as little machinery as possible to non-compiler tools.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks! (And apologies for the extreme delay here. :scream_cat:) I've pushed changes to use the new STL error message machinery, to perform this check for only real compilers (non-compiler tools have "preprocessors" with sometimes questionable behavior), and to fuse the test coverage into the pre-existing test for core headers where this makes perfect sense.

Users will see:

```
D:\GitHub\STL>type meow.cpp
#include <vector>
int main() {}

D:\GitHub\STL>cl /EHsc /nologo /W4 /D_ENFORCE_ONLY_CORE_HEADERS meow.cpp
meow.cpp
D:\GitHub\STL\out\build\x64\out\inc\yvals.h(14): error STL1005: Tried to include a non-core C++ Standard Library header file with _ENFORCE_ONLY_CORE_HEADERS defined.
D:\GitHub\STL\out\build\x64\out\inc\yvals.h(13): error C2338: static_assert failed: 'Error in C++ Standard Library usage.'
```

**Review by CaseyCarter — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks for making it easier for users to limit themselves to the core subset! :dart: :white_check_mark: :tada:
